In [84]:
import numpy as np
import pandas as pd
import itertools as it
import matplotlib
import matplotlib.pyplot as plt
import matplotlib.cm as cmx
import matplotlib.colors as colors
import scipy as sp
import statsmodels.api as sm
from itertools import combinations
from sklearn import preprocessing
from sklearn.cross_validation import train_test_split
from sklearn.linear_model import LinearRegression
from collections import Counter
import sys
%matplotlib inline

In [30]:
#read the data
df = pd.read_csv('listings_clean.csv')
print(df.columns.values)
df.head()

['id' 'host_since' 'zipcode' 'latitude' 'longitude' 'property_type'
 'room_type' 'accommodates' 'bathrooms' 'bedrooms' 'beds' 'bed_type'
 'guests_included' 'minimum_nights' 'maximum_nights' 'availability_30'
 'availability_60' 'availability_90' 'availability_365' 'number_of_reviews'
 'first_review' 'last_review' 'review_scores_rating'
 'review_scores_accuracy' 'review_scores_cleanliness'
 'review_scores_checkin' 'review_scores_communication'
 'review_scores_location' 'review_scores_value' 'host_listing_count'
 'price']


id  host_since  zipcode   latitude  longitude  property_type  \
0  1069266        1926  10022.0  40.756852 -73.964754              0   
1  1846722        1625      NaN  40.830599 -73.941014              0   
2  2061725        1831  11221.0  40.692189 -73.924120              0   
3    44974         953  10011.0  40.734751 -74.002592              0   
4  4701675        2479  10011.0  40.745282 -73.997836              0   

   room_type  accommodates  bathrooms  bedrooms  ...    last_review  \
0          0             2        1.0       1.0  ...         2542.0   
1          0            10        1.0       3.0  ...         2554.0   
2          1             2        1.0       1.0  ...         2554.0   
3          0             2        1.0       1.0  ...         2494.0   
4          0             2        1.0       1.0  ...         2533.0   

   review_scores_rating  review_scores_accuracy  review_scores_cleanliness  \
0                  86.0                     9.0                        7.0   
1                  85.0                     8.0                        8.0   
2                  98.0                    10.0                       10.0   
3                  96.0                    10.0                        9.0   
4                 100.0                    10.0                       10.0   

   review_scores_checkin  review_scores_communication  review_scores_location  \
0                    9.0                          9.0                    10.0   
1                    9.0                          8.0                     7.0   
2                   10.0                         10.0                     9.0   
3                   10.0                         10.0                    10.0   
4                   10.0                         10.0                    10.0   

   review_scores_value  host_listing_count  price  
0                  9.0                   1    160  
1                  8.0                   2    105  
2                 10.0                   4     58  
3                  9.0                   1    185  
4                 10.0                   1    195  

[5 rows x 31 columns]

In [31]:
df.shape

(27392, 31)

In [2]:
##Added Calendar Information to OLS model
#read calendar data
data = pd.read_csv("calendar_grid.csv")
print data.shape
data.head(n=20)

(27392, 366)


listing_id  01-01  01-02  01-03  01-04  01-05  01-06  01-07  01-08  01-09  \
0      3604481  600.0  600.0  600.0  600.0  600.0  600.0  600.0  600.0  600.0   
1      2949128    NaN    NaN    NaN    NaN    NaN  100.0  100.0  100.0  100.0   
2      4325397  100.0  100.0  100.0  100.0  100.0   70.0   70.0   70.0   80.0   
3      4325398    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4      3426149    NaN    NaN    NaN    NaN  100.0  100.0    NaN    NaN    NaN   
5        65562    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
6      2730672   58.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
7      4587554    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
8       819206    NaN    NaN    NaN   70.0   70.0   70.0   70.0   70.0   70.0   
9      4849703    NaN  100.0  100.0    NaN    NaN    NaN    NaN    NaN    NaN   
10     2359340    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
11     4653104  150.0  150.0  150.0    NaN    NaN    NaN    NaN    NaN    NaN   
12     3276860    NaN    NaN    NaN  450.0  450.0  375.0  375.0  375.0  375.0   
13     3604551  150.0  150.0  150.0  150.0  100.0  100.0  100.0  100.0  100.0   
14     3997768    NaN   42.0   42.0   42.0   42.0    NaN    NaN    NaN    NaN   
15     4128844    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  130.0   
16      185698    NaN    NaN    NaN   50.0   50.0   50.0   50.0    NaN    NaN   
17     3866715  350.0  320.0  320.0  320.0  320.0  320.0  320.0  320.0  320.0   
18     2752604    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
19      262238    NaN    NaN  168.0  168.0  168.0  168.0  168.0  168.0  168.0   

    ...    12-22  12-23  12-24  12-25  12-26  12-27  12-28  12-29  12-30  \
0   ...    600.0  600.0  600.0  600.0  600.0  600.0  600.0  600.0  600.0   
1   ...    100.0  100.0  100.0  100.0  100.0  100.0  100.0  100.0  100.0   
2   ...     70.0   70.0   70.0   80.0   80.0   70.0   70.0   70.0   70.0   
3   ...    275.0  275.0  275.0  275.0  275.0  275.0  275.0  275.0  275.0   
4   ...     95.0   95.0   95.0   95.0   95.0   95.0   95.0   95.0   95.0   
5   ...     97.0   97.0   97.0   97.0   97.0   97.0   97.0   97.0   97.0   
6   ...     58.0   58.0   58.0   58.0   58.0   58.0   58.0   58.0   58.0   
7   ...     55.0   55.0   55.0   55.0   55.0   55.0   55.0   55.0   55.0   
8   ...     70.0   70.0   70.0   70.0   70.0   70.0   70.0   70.0   70.0   
9   ...      NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
10  ...    235.0  235.0  425.0  249.0  249.0  235.0  235.0  235.0  235.0   
11  ...    150.0  150.0  150.0  150.0  150.0  150.0  150.0  150.0  150.0   
12  ...    450.0  450.0  450.0  450.0  450.0  450.0  450.0  450.0  450.0   
13  ...    100.0  100.0  100.0  100.0  100.0  100.0  100.0  100.0  100.0   
14  ...     49.0   49.0   49.0   49.0   49.0   49.0   49.0   49.0   49.0   
15  ...    100.0  100.0  100.0  130.0  130.0  100.0  100.0  100.0  100.0   
16  ...     50.0   50.0   50.0   50.0   50.0   50.0   50.0   50.0   50.0   
17  ...    320.0  320.0  320.0  320.0  320.0  320.0  320.0  320.0  320.0   
18  ...    100.0  100.0  100.0  100.0  100.0  100.0  100.0  100.0  100.0   
19  ...    168.0  168.0  168.0  168.0  168.0  168.0  168.0  168.0  168.0   

    12-31  
0   600.0  
1   100.0  
2    70.0  
3   275.0  
4    95.0  
5    97.0  
6    58.0  
7    55.0  
8    70.0  
9     NaN  
10  479.0  
11  150.0  
12  450.0  
13  100.0  
14   49.0  
15  100.0  
16   50.0  
17  320.0  
18  100.0  
19  168.0  

[20 rows x 366 columns]

In [26]:
#adding 2 binary indicator variable columns and duplicating the listing data 4 times for each combination.
#2015.7.4 is Saturday
#So we need to extract: 
    #                                       Holiday     Weekend
    #        7.4                             yes          yes
    #  1.1, 11.26, 12.25                     yes          no
    #   all weekends-7.4                     no           yes
    #   all weekdays- 1.1, 11.26, 12.25      no           no
    
#holiday indices
holiday_dates = ['01-01', '07-04', '11-26', '12-25']
holiday_index = [data.columns.get_loc(x) for x in holiday_dates]

weekend_holiday_index = [i for i in range(1, 366) if (i % 7 == 3 or i % 7 == 4) and (i in holiday_index)]
weekend_nonholiday_index = [i for i in range(1, 366) if (i % 7 == 3 or i % 7 == 4) and (i not in holiday_index)]
weekday_holiday_index = [i for i in range(1, 366) if (i % 7 < 3 or i % 7 > 4) and (i in holiday_index)]
weekday_nonholiday_index = [i for i in range(1, 366) if (i % 7 < 3 or i % 7 > 4) and (i not in holiday_index)]


#get averages
weekend_holidays = data.values[:, weekend_holiday_index]
weekend_nonholidays = data.values[:, weekend_nonholiday_index]
weekday_holidays = data.values[:, weekday_holiday_index]
weekday_nonholidays = data.values[:, weekday_nonholiday_index]

weekend_holiday_prices = np.nanmean(weekend_holidays, axis=1)
weekend_nonholiday_prices = np.nanmean(weekend_nonholidays, axis=1)
weekday_holiday_prices = np.nanmean(weekday_holidays, axis=1)
weekday_nonholiday_prices = np.nanmean(weekday_nonholidays, axis=1)

In [29]:
print weekend_holiday_prices.shape
print weekend_nonholiday_prices.shape
print weekday_holiday_prices.shape
print weekday_nonholiday_prices.shape

(27392,)
(27392,)
(27392,)
(27392,)


In [87]:
#Modify df using calendar infos
print df.shape
features = np.append(df.columns.values, ['weekend', 'holiday'])
new_df = pd.DataFrame(np.zeros((df.shape[0]*4, df.shape[1]+2)), columns=features)
print new_df.shape

(27392, 31)
(109568, 33)


In [89]:
#now append items
#duplicate each original entry for four times

for i in range(df.shape[0]):
    #find the index in calendar file
    #duplicate four times and append to new_df    
    item = df.iloc[i].copy()
    item['weekend']=1

    item['holiday']=1
    item['price']=weekend_holiday_prices[data[data['listing_id']==df.iloc[i]['id']].index.tolist()[0]]
    new_df.iloc[i*4]=item
    item['holiday']=0
    item['price']=weekend_nonholiday_prices[data[data['listing_id']==df.iloc[i]['id']].index.tolist()[0]]
    new_df.iloc[i*4+1]=item

    item['weekend']=0

    item['holiday']=1
    item['price']=weekday_holiday_prices[data[data['listing_id']==df.iloc[i]['id']].index.tolist()[0]]
    new_df.iloc[i*4+2]=item
    item['holiday']=0
    item['price']=weekday_nonholiday_prices[data[data['listing_id']==df.iloc[i]['id']].index.tolist()[0]]
    new_df.iloc[i*4+3]=item
    
    sys.stdout.write("\r{0}% completed".format((float(i+1)/df.shape[0])*100))
    sys.stdout.flush()

100.0% completed

In [90]:
new_df

id  host_since  zipcode   latitude  longitude  property_type  \
0       1069266.0      1926.0  10022.0  40.756852 -73.964754            0.0   
1       1069266.0      1926.0  10022.0  40.756852 -73.964754            0.0   
2       1069266.0      1926.0  10022.0  40.756852 -73.964754            0.0   
3       1069266.0      1926.0  10022.0  40.756852 -73.964754            0.0   
4       1846722.0      1625.0      NaN  40.830599 -73.941014            0.0   
5       1846722.0      1625.0      NaN  40.830599 -73.941014            0.0   
6       1846722.0      1625.0      NaN  40.830599 -73.941014            0.0   
7       1846722.0      1625.0      NaN  40.830599 -73.941014            0.0   
8       2061725.0      1831.0  11221.0  40.692189 -73.924120            0.0   
9       2061725.0      1831.0  11221.0  40.692189 -73.924120            0.0   
10      2061725.0      1831.0  11221.0  40.692189 -73.924120            0.0   
11      2061725.0      1831.0  11221.0  40.692189 -73.924120            0.0   
12        44974.0       953.0  10011.0  40.734751 -74.002592            0.0   
13        44974.0       953.0  10011.0  40.734751 -74.002592            0.0   
14        44974.0       953.0  10011.0  40.734751 -74.002592            0.0   
15        44974.0       953.0  10011.0  40.734751 -74.002592            0.0   
16      4701675.0      2479.0  10011.0  40.745282 -73.997836            0.0   
17      4701675.0      2479.0  10011.0  40.745282 -73.997836            0.0   
18      4701675.0      2479.0  10011.0  40.745282 -73.997836            0.0   
19      4701675.0      2479.0  10011.0  40.745282 -73.997836            0.0   
20        68914.0      1106.0  11231.0  40.679060 -73.994730            0.0   
21        68914.0      1106.0  11231.0  40.679060 -73.994730            0.0   
22        68914.0      1106.0  11231.0  40.679060 -73.994730            0.0   
23        68914.0      1106.0  11231.0  40.679060 -73.994730            0.0   
24      4832596.0      1779.0  11207.0  40.684422 -73.909882            0.0   
25      4832596.0      1779.0  11207.0  40.684422 -73.909882            0.0   
26      4832596.0      1779.0  11207.0  40.684422 -73.909882            0.0   
27      4832596.0      1779.0  11207.0  40.684422 -73.909882            0.0   
28      2562510.0      2264.0  10013.0  40.719319 -73.997247            0.0   
29      2562510.0      2264.0  10013.0  40.719319 -73.997247            0.0   
...           ...         ...      ...        ...        ...            ...   
109538   499763.0      1120.0  10019.0  40.764509 -73.988566            0.0   
109539   499763.0      1120.0  10019.0  40.764509 -73.988566            0.0   
109540  2361020.0      1904.0  11233.0  40.681716 -73.932360            0.0   
109541  2361020.0      1904.0  11233.0  40.681716 -73.932360            0.0   
109542  2361020.0      1904.0  11233.0  40.681716 -73.932360            0.0   
109543  2361020.0      1904.0  11233.0  40.681716 -73.932360            0.0   
109544   659875.0      1538.0  10003.0  40.726843 -73.989648            0.0   
109545   659875.0      1538.0  10003.0  40.726843 -73.989648            0.0   
109546   659875.0      1538.0  10003.0  40.726843 -73.989648            0.0   
109547   659875.0      1538.0  10003.0  40.726843 -73.989648            0.0   
109548   661849.0      1437.0  10016.0  40.742363 -73.981968            0.0   
109549   661849.0      1437.0  10016.0  40.742363 -73.981968            0.0   
109550   661849.0      1437.0  10016.0  40.742363 -73.981968            0.0   
109551   661849.0      1437.0  10016.0  40.742363 -73.981968            0.0   
109552  1080766.0      1931.0  10019.0  40.759642 -73.985343            0.0   
109553  1080766.0      1931.0  10019.0  40.759642 -73.985343            0.0   
109554  1080766.0      1931.0  10019.0  40.759642 -73.985343            0.0   
109555  1080766.0      1931.0  10019.0  40.759642 -73.985343            0.0   
109556  2668770.0      1630.0  10301.0  40.645741 -74.080955     

In [93]:
# remove small zipcode clusters, with 95% data left
# count each zipcode entries
nb_counts = Counter(new_df['zipcode'])
tdf = pd.DataFrame.from_dict(nb_counts, orient='index').sort_values(by=0)
# select clusters with size >= 50
tdf1 = tdf[tdf.values>=50*4]
print 'Remaining data proportion: ' + str(float(sum(tdf1.values))/new_df.shape[0])
# get zipcode clusters
zipcode_included = tdf1.index

new_df1 = new_df.loc[new_df['zipcode'].isin(zipcode_included)]
print new_df1.shape

Remaining data proportion: 0.94976635514
(104064, 33)


In [117]:
#remove items that price doesn't exist
new_df2 = new_df1[~np.isnan(new_df1['price'])]

print 'Remaining data proportion: ' + str(float(new_df2.shape[0])/new_df.shape[0])
print new_df2.shape

Remaining data proportion: 0.847136025117
(92819, 33)


In [118]:
new_y = new_df2['price']

In [119]:
#Feature Extraction:
#Use the features we decided last time
features = ['zipcode', 'latitude', 'longitude', 'property_type', 'room_type', 'accommodates', 
            'bathrooms', 'bedrooms', 'beds', 'bed_type', 'number_of_reviews', 'review_scores_rating',
            'review_scores_accuracy', 'review_scores_cleanliness', 
            'review_scores_checkin', 'review_scores_communication', 
            'review_scores_location', 'review_scores_value', 'host_listing_count']

new_df_x = new_df2[features]
new_df_x.head()

zipcode   latitude  longitude  property_type  room_type  accommodates  \
0  10022.0  40.756852 -73.964754            0.0        0.0           2.0   
1  10022.0  40.756852 -73.964754            0.0        0.0           2.0   
2  10022.0  40.756852 -73.964754            0.0        0.0           2.0   
3  10022.0  40.756852 -73.964754            0.0        0.0           2.0   
9  11221.0  40.692189 -73.924120            0.0        1.0           2.0   

   bathrooms  bedrooms  beds  bed_type  number_of_reviews  \
0        1.0       1.0   1.0       0.0               62.0   
1        1.0       1.0   1.0       0.0               62.0   
2        1.0       1.0   1.0       0.0               62.0   
3        1.0       1.0   1.0       0.0               62.0   
9        1.0       1.0   2.0       0.0               35.0   

   review_scores_rating  review_scores_accuracy  review_scores_cleanliness  \
0                  86.0                     9.0                        7.0   
1                  86.0                     9.0                        7.0   
2                  86.0                     9.0                        7.0   
3                  86.0                     9.0                        7.0   
9                  98.0                    10.0                       10.0   

   review_scores_checkin  review_scores_communication  review_scores_location  \
0                    9.0                          9.0                    10.0   
1                    9.0                          9.0                    10.0   
2                    9.0                          9.0                    10.0   
3                    9.0                          9.0                    10.0   
9                   10.0                         10.0                     9.0   

   review_scores_value  host_listing_count  
0                  9.0                 1.0  
1                  9.0                 1.0  
2                  9.0                 1.0  
3                  9.0                 1.0  
9                 10.0                 4.0

In [120]:
#categorical and numerical column names
cate = ['zipcode', 'property_type', 'room_type', 'bed_type']
nume = [c for c in new_df_x.columns.values if c not in cate]

#fill categorical with mode, numerical with mean
new_df_x_cate = new_df_x[cate].apply(lambda x:x.fillna(x.value_counts().index[0]))
new_df_x_nume = new_df_x[nume].apply(lambda x:x.fillna(x.mean()))

new_df_x = pd.concat([new_df_x_cate, new_df_x_nume], axis=1)
new_df_x.head()

zipcode  property_type  room_type  bed_type   latitude  longitude  \
0  10022.0            0.0        0.0       0.0  40.756852 -73.964754   
1  10022.0            0.0        0.0       0.0  40.756852 -73.964754   
2  10022.0            0.0        0.0       0.0  40.756852 -73.964754   
3  10022.0            0.0        0.0       0.0  40.756852 -73.964754   
9  11221.0            0.0        1.0       0.0  40.692189 -73.924120   

   accommodates  bathrooms  bedrooms  beds  number_of_reviews  \
0           2.0        1.0       1.0   1.0               62.0   
1           2.0        1.0       1.0   1.0               62.0   
2           2.0        1.0       1.0   1.0               62.0   
3           2.0        1.0       1.0   1.0               62.0   
9           2.0        1.0       1.0   2.0               35.0   

   review_scores_rating  review_scores_accuracy  review_scores_cleanliness  \
0                  86.0                     9.0                        7.0   
1                  86.0                     9.0                        7.0   
2                  86.0                     9.0                        7.0   
3                  86.0                     9.0                        7.0   
9                  98.0                    10.0                       10.0   

   review_scores_checkin  review_scores_communication  review_scores_location  \
0                    9.0                          9.0                    10.0   
1                    9.0                          9.0                    10.0   
2                    9.0                          9.0                    10.0   
3                    9.0                          9.0                    10.0   
9                   10.0                         10.0                     9.0   

   review_scores_value  host_listing_count  
0                  9.0                 1.0  
1                  9.0                 1.0  
2                  9.0                 1.0  
3                  9.0                 1.0  
9                 10.0                 4.0

In [122]:
# Get numpy array from data
new_x = new_df_x.values

# Apply one hot endcoing
categorical = [(c in cate) for c in new_df_x.columns]
encoder = preprocessing.OneHotEncoder(categorical_features=categorical, sparse=False)  

new_x = encoder.fit_transform(new_x)

print new_x.shape
print new_y.shape

(92819, 103)
(92819,)


In [123]:
x_train, x_test, y_train, y_test = train_test_split(new_x, new_y, test_size=0.3, random_state=0)
print 'train data: ', x_train.shape
print 'test data: ', x_test.shape
#OLS
X = sm.add_constant(x_train)
model = sm.OLS(y_train,X)
results = model.fit()
print results.summary()

train data:  (64973, 103)
test data:  (27846, 103)
                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.453
Model:                            OLS   Adj. R-squared:                  0.452
Method:                 Least Squares   F-statistic:                     542.9
Date:                Tue, 06 Dec 2016   Prob (F-statistic):               0.00
Time:                        18:24:02   Log-Likelihood:            -3.8082e+05
No. Observations:               64973   AIC:                         7.618e+05
Df Residuals:                   64873   BIC:                         7.628e+05
Df Model:                          99                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
c

In [124]:
#same OLS model with sklearn (easier to score but less detailed output)
model = LinearRegression()
model.fit(x_train, y_train)
print "Train Score", model.score(x_train, y_train)
print "Test Score", model.score(x_test, y_test)

Train Score 0.453104355163
Test Score 0.45388020956


In [126]:
#try several random seeds to check for stability and variance
for r in range(10):
    x_train, x_test, y_train, y_test = train_test_split(new_x, new_y, test_size=0.3, random_state=r)
    model = LinearRegression()
    model.fit(x_train, y_train)
    print "Train {:.3f} Test {:.3f}".format(model.score(x_train, y_train), model.score(x_test, y_test))

Train 0.453 Test 0.454
Train 0.454 Test 0.452
Train 0.453 Test 0.455
Train 0.454 Test 0.453
Train 0.459 Test 0.441
Train 0.457 Test 0.444
Train 0.458 Test 0.442
Train 0.454 Test 0.451
Train 0.453 Test 0.454
Train 0.453 Test 0.455


We could see that after encoding the calendar information, we now could get a score of 0.45 using OLS.